src/fu-smbios.c

```diff
@@ -0,0 +1,378 @@
+		g_string_append_printf (str, " Handle: 0x%04x\n", item->handle);
+		for (guint j = 0; j < item->strings->len; j++) {
+			const gchar *tmp = g_ptr_array_index (item->strings, j);
+			g_string_append_printf (str, "  String[%02u]: %s\n", j, tmp);
+		}
+	}
+	return g_string_free (str, FALSE);
+}
+
+static FuSmbiosItem *
+fu_smbios_get_item_for_type (FuSmbios *self, guint8 type)
+{
+	for (guint i = 0; i < self->items->len; i++) {
+		FuSmbiosItem *item = g_ptr_array_index (self->items, i);
+		if (item->type == type)
+			return item;
+	}
+	return NULL;
+}
+
+/**
+ * fu_smbios_get_data:
+ * @self: A #FuSmbios
+ * @type: A structure type
+ * @error: A #GError or %NULL
+ *
+ * Reads a SMBIOS data blob, which includes the SMBIOS section header.
+ *
+ * Returns: a #GBytes, or %NULL if invalid or not found
+ **/
+GBytes *
+fu_smbios_get_data (FuSmbios *self, guint8 type, GError **error)
+{
+	FuSmbiosItem *item;
+	g_return_val_if_fail (FU_IS_SMBIOS (self), NULL);
+	item = fu_smbios_get_item_for_type (self, type);
+	if (item == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "no structure with type %02x", type);
+		return NULL;
+	}
+	return item->data;
+}
+
+/**
+ * fu_smbios_get_string:
+ * @self: A #FuSmbios
+ * @type: A structure type
+ * @offset: A structure offset
+ * @error: A #GError or %NULL
+ *
+ * Reads a string from the SMBIOS string table of a specific structure.
+ *
+ * Returns: a string, or %NULL if invalid or not found
+ **/
+const gchar *
+fu_smbios_get_string (FuSmbios *self, guint8 type, guint8 offset, GError **error)
+{
+	FuSmbiosItem *item;
+	const guint8 *data;
+	gsize sz;
+
+	g_return_val_if_fail (FU_IS_SMBIOS (self), NULL);
+
+	/* get item */
+	item = fu_smbios_get_item_for_type (self, type);
+	if (item == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "no structure with type %02x", type);
+		return NULL;
+	}
+
+	/* check offset valid */
+	data = g_bytes_get_data (item->data, &sz);
+	if (offset >= sz) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "offset bigger than size %" G_GSIZE_FORMAT, sz);
+		return NULL;
+	}
+	if (data[offset] == 0x00) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_FOUND,
+			     "no data available");
+		return NULL;
+	}
+
+	/* check string index valid */
+	if (data[offset] > item->strings->len) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INVALID_FILE,
+			     "index larger than string table %u",
+			     data[offset]);
+		return NULL;
+	}
+	return g_ptr_array_index (item->strings, data[offset] - 1);
+}
+
+static void
+fu_smbios_item_free (FuSmbiosItem *item)
+{
+	g_bytes_unref (item->data);
+	g_ptr_array_unref (item->strings);
+	g_free (item);
+}
+
+static void
+fu_smbios_finalize (GObject *object)
+{
+	FuSmbios *self;
+	g_return_if_fail (FU_IS_SMBIOS (object));
+	self = FU_SMBIOS (object);
```

Comment:
Is there a situation you were envisioning this g_return_if_fail is necessary during finalize?  Some failed initialization or something?

Reply:
Agreed, I've dropped this. At some point it appeared and I've C&P'ed ever since.